internal/elasticsearch/security/user/update.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+func (r *userResource) update(ctx context.Context, plan tfsdk.Plan, state *tfsdk.State) diag.Diagnostics {
+	var planData UserData
+	var diags diag.Diagnostics
+	diags.Append(plan.Get(ctx, &planData)...)
+	if diags.HasError() {
+		return diags
+	}
+
+	// Check if we have existing state (this is an update, not a create)
+	hasState := false
+	var stateData UserData
+	if state != nil {
+		if getRaw := state.Raw; !getRaw.IsNull() {
+			hasState = true
+			diags.Append(state.Get(ctx, &stateData)...)
+			if diags.HasError() {
+				return diags
+			}
+		}
+	}
+
+	usernameId := planData.Username.ValueString()
+	id, sdkDiags := r.client.ID(ctx, usernameId)
+	diags.Append(diagutil.FrameworkDiagsFromSDK(sdkDiags)...)
+	if diags.HasError() {
+		return diags
+	}
+
+	client, connDiags := clients.MaybeNewApiClientFromFrameworkResource(ctx, planData.ElasticsearchConnection, r.client)
+	diags.Append(connDiags...)
+	if diags.HasError() {
+		return diags
+	}
+
+	var user models.User
+	user.Username = usernameId
+
+	// Handle password fields - only set password if it's in the plan AND (it's a create OR it has changed from state)
+	// Priority: password_wo > password > password_hash
+	if utils.IsKnown(planData.PasswordWo) && (!hasState || !planData.PasswordWo.Equal(stateData.PasswordWo) || !planData.PasswordWoVersion.Equal(stateData.PasswordWoVersion)) {
+		// Use write-only password
+		password := planData.PasswordWo.ValueString()
+		user.Password = &password
+	} else if utils.IsKnown(planData.Password) && (!hasState || !planData.Password.Equal(stateData.Password)) {
+		password := planData.Password.ValueString()
+		user.Password = &password
+	} else if utils.IsKnown(planData.PasswordHash) && (!hasState || !planData.PasswordHash.Equal(stateData.PasswordHash)) {
+		passwordHash := planData.PasswordHash.ValueString()
+		user.PasswordHash = &passwordHash
+	}
+
+	if utils.IsKnown(planData.Email) {
+		user.Email = planData.Email.ValueString()
+	}
+	if utils.IsKnown(planData.FullName) {
+		user.FullName = planData.FullName.ValueString()
+	}
+	user.Enabled = planData.Enabled.ValueBool()
+
+	roles := make([]string, 0, len(planData.Roles.Elements()))
+	diags.Append(planData.Roles.ElementsAs(ctx, &roles, false)...)
+	if diags.HasError() {
+		return diags
+	}
+	user.Roles = roles
+
+	if !planData.Metadata.IsNull() && !planData.Metadata.IsUnknown() {
+		metadata := make(map[string]interface{})
+		if err := json.NewDecoder(strings.NewReader(planData.Metadata.ValueString())).Decode(&metadata); err != nil {
+			diags.AddError("Failed to decode metadata", err.Error())
+			return diags
+		}
+		user.Metadata = metadata
+	}
+
+	sdkDiags = elasticsearch.PutUser(ctx, client, &user)
```

Comment:
This `PutUser` function is only ever called from Plugin Framework code. Update the function to return framework diagnostics so we don't have to do the translation here.